### PR TITLE
[WIP] Improved local testing

### PR DESCRIPTION
**Pull request by SamMousa**
The goal of this PR is to improve the experience of Yii2 contributors.
It does this by providing a `docker-compose` file that loads all dependencies and runs the tests.

This branch also contains several improvements to test speed as well as a phpunit plugin that lists slow tests.

| Q             | A
| ------------- | ---
| Is bugfix?    | yes/no
| New feature?  | yes/no
| Breaks BC?    | yes/no
| Tests pass?   | yes/no
| Fixed issues  | comma-separated list of tickets # fixed by the PR, if any

### CODE REVIEW

**Comment by schmunk42** [comment]
There's already a [Dockerfile](https://github.com/yiisoft/yii2/blob/master/Dockerfile) along with `docker-compose` files in `tests/`. We should improve the existing ones, if needed. They are also based on Yii's Docker images, which are based on the official PHP ones.

**Comment by SamMousa** [comment]
I found those files and none of them really do what I need them to do; as far as I can tell they are indeed based on the official PHP docker images but that is all.
For every purpose (even every yii application) you should have your own docker file to install relevant php extensions and other dependencies.
For development it makes sense to have all possible dependencies so you can run all tests.

@schmunk42 Which docker-compose file specifically would you use for local testing? I tried some of them and they don't really seem to do what I need as a developer...

**Comment by schmunk42** [comment]
> For every purpose (even every yii application) you should have your own docker file to install relevant php extensions and other dependencies.

There's a quite large set which is supported by the "official" base-images, in fact all extensions checked by the requirement checker, including MySQL and Postgre, but not Cubrid, OCI and MSSQL, along with some Windows specific caching. Let me know what you are missing.

> For every purpose (even every yii application) you should have your own docker file to install relevant php extensions and other dependencies.

Here is some info about [installing additional extensions](https://github.com/yiisoft/yii2-docker/blob/master/docs/install-extensions.md). 

> For development it makes sense to have all possible dependencies so you can run all tests.

Might make sense to add some drivers for testing... but it should be done on top of the "official" images.

> Which docker-compose file specifically would you use for local testing? I tried some of them and they don't really seem to do what I need as a developer...

I've added a description about how to use the multiple `docker-compose` files, please see https://github.com/yiisoft/yii2/pull/15346

**Comment by SamMousa** [comment]
@schmunk42 to me it all looks quite complicated for something that should be a simple one liner.

I'm not sure why I would want to run 2 commands (`up` and `run`) when docker compose can figure that out for itself.

Docker makes it possible to get started with local development doing something like this:
````
git clone https://github.com/yiisoft/yii2.git && cd yii2
composer install
docker-compose -f tests/docker-compose-development.yml run phpunit
````

> Might make sense to add some drivers for testing... but it should be done on top of the "official" images.

Why? What do these official images do other than install PHP and it's dependencies in an efficient way:
````
local/yiisoftware/yii2-php                              7.1.9-fpm-alpine    a900ae3049e3        About a minute ago   240MB
tests_phpunit                                           latest              4879ef0a7eb3        4 hours ago          65MB
tests_shell                                             latest              4879ef0a7eb3        4 hours ago          65MB
````
I don't get why you would base your image on the official image for deployment. Every package you add but don't use increases your attack surface, offering an official image will mean people will tend to use it as a base `FROM yiisoftware/yii2` and thus install things they don't need and that could in theory even hurt them.

### Side note:
I've looked at yii2-docker and I don't like the way it works. (In the example compose files)
In my opinion, for production, one should not mount the PHP into docker at all.
I've created and am using an nginx container that does not contain the PHP source code via https://github.com/sam-it/yii2-static-assets
Further I'm building my php-fpm container without mounting source code into it.

The mounting approach is great for development, but for production it comes with a lot of issues, especially when you start scaling up and want to use multiple fpm containers with load balancing.

File: tests/docker-compose.development.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      - redis
+      - memcached
+      - mysql
+      - postgres
+    volumes:
+      - ..:/project
+      - ./data/config-docker.php:/project/tests/data/config.php
+    entrypoint:
+      - /bin/sh
+  phpunit:
```
Comment (schmunk42) [inline tests/docker-compose.development.yml]: What would be the difference between `shell` and `phpunit`?

Comment (SamMousa) [reply tests/docker-compose.development.yml]: Just the entrypoint. Phpunit runs phpunit, so yo ucan run it like `docker-compose run phpunit --list-groups`

**Comment by SamMousa** [comment]
> It's merely due to the fact that databases might need a few seconds until they are fully ready even if the PHP container depends on the databases. So tests might run before databases are completely initialized.

Ah, that's a very valid point several solutions exist for it, here's what I use (for a random private project):
````
codeception:
    depends_on:
      - mysql
      - redis
    tty: true
    build: ./dockerfiles/codeception
    entrypoint: [
      "/sbin/tini",
      "/bin/wait-for-it",
      "-h",
      "$$DB_HOST:3306",
      "--",
      "/project/bin/codecept"
    ]
    volumes:
      - ./:/project
    environment:
      YII_ENV: test
      DB_HOST: mysql
      DB_USER: root
      DB_PASS: secret
      DB_NAME: cpm
      REDIS_HOST: redis
      REDIS_PASS: secret
````
Where `wait-for-it` is something I bake into the test image. (https://github.com/SAM-IT/wait-for-it-php)

> PHP + nginx + assets are a pain, I know. We added nginx to the PHP image for this reason, but it's nothing I'd consider making a default.

Ah, I guess at the current stage it needs some more maturing then.

Anyway, in conclusion I think the developer should get a simple solution that Just Works and if he or she wants to optimize the environment they can. Currently it's the other way around and getting tests to run requires several steps. That's what I'm trying to combat here.

Also note I'm including `tini`, an init that allows me to properly interrupt test using `ctrl + c`. (It is build in to docker nowadays, and on my todo to use: https://docs.docker.com/compose/compose-file/compose-file-v2/#init)

**Comment by schmunk42** [comment]
> Where wait-for-it is something I bake into the test image. (https://github.com/SAM-IT/wait-for-it-php)

more for reference ... I did this for gitlab tests with those "one-liners", a bit quirky - but works. For local testing I'd recommend `docker-compose up -d`

    docker-compose run --rm php bash -c "while ! curl mysql:3306; do ((c++)) && ((c==30)) && break; sleep 2; done"
    docker-compose run --rm php bash -c 'while [ true ]; do curl postgres:5432; if [ $? == 52 ]; then break; fi; ((c++)) && ((c==25)) && break; sleep 2; done'

https://github.com/yiisoft/yii2/blob/master/.gitlab-ci.yml#L56

> Anyway, in conclusion I think the developer should get a simple solution that Just Works and if he or she wants to optimize the environment they can. Currently it's the other way around and getting tests to run requires several steps. That's what I'm trying to combat here.

WDYT about https://github.com/yiisoft/yii2/pull/15346/files#r156429478 ? - if we would uncomment this line you could run the majority of test with.

```
cd tests
cp .env-dist .env
docker-compose up -d
docker-compose run --rm php vendor/bin/phpunit -v 
```

**Review by schmunk42 — CHANGES_REQUESTED** [review]
With the current change in docs, all tests should run out-of-the box with the current Docker setup.

I think `Dockerfile` and `docker-compose.development.yml` should be removed from this PR.

**Comment by samdark** [comment]
Doesn't seem to be valid/active anymore. Closing.

**Pull request closed without merging**